File: C++/word-break.cpp
class Solution {
public:
    bool wordBreak(std::string s, unordered_set<std::string>& wordDict) {
        int n = s.size();
        bool dp[n] = {false};
        for(int i=n-1; i>-1; --i) {
            if(wordDict.find(s.substr(i, n))!=wordDict.end()) {
                dp[i] = true;
            }
            else {
                for(int j=1; j<n; ++j) {
                    if(dp[j] && wordDict.find(s.substr(i, j-i))!=wordDict.end()) {
                        dp[i] = true;
                        break;
                    }
                }
            }
        }
        return dp[0];
    }
};

File: Python/word-break.py
class Solution:
    # @param s, a string
    # @param wordDict, a set<string>
    # @return a boolean
    def wordBreak(self, s, wordDict):
        n = len(s)
        dp = [False]*n
        for i in xrange(n-1,-1,-1):
            if s[i:n] in wordDict:
                dp[i] = True
            else:
                for j in xrange(i+1, n):
                    if dp[j] and s[i:j] in wordDict:
                        dp[i] = True
                        break
        return dp[0]